tensorflow/compiler/xla/service/slice_delaying.cc

```diff
@@ -0,0 +1,295 @@
+          "Operation's true operand should be the same shape");
+    }
+  }
+  return operands;
+}
+
+StatusOr<std::vector<HloInstruction*>> SliceDelayer::GetTrueUsers(
+    const HloInstruction* inst,
+    const std::vector<HloInstruction*>& operands) {
+  std::vector<HloInstruction*> users;
+  HloInstruction* operand0 = operands[0];
+
+  for (const HloInstruction* slice_0 : operand0->users()) {
+    // skip non-slice user
+    if (slice_0->opcode() != HloOpcode::kSlice) {
+      continue;
+    }
+
+    for (HloInstruction* user : slice_0->users()) {
+      // user should be the same operation and same operand count as inst
+      // skip the visited user to avoid redundant computation
+      if (IsVisited(user) || user->opcode() != inst->opcode() ||
+          user->operand_count() != inst->operand_count()) {
+        continue;
+      }
+
+      // user's every operand should be a slice of the true operand in order
+      bool isValidUser = true;
+      for (int64 j = 0; j < operands.size(); ++j) {
+        const HloInstruction* slice_j = user->operand(j);
+        // the slice of operands should sliced from the same location (only for
+        // elementwise)
+        if (slice_j->opcode() != HloOpcode::kSlice ||
+            slice_j->operand(0) != operands[j] ||
+            slice_0->slice_starts() != slice_j->slice_starts() ||
+            slice_0->slice_limits() != slice_j->slice_limits() ||
+            slice_0->slice_strides() != slice_j->slice_strides()) {
+          isValidUser = false;
+          break;
+        }
+      }
+      if (!isValidUser) {
+        continue;
+      }
+
+      // found the user
+      users.push_back(user);
+      visited_.insert(user);
+    }  // end for loop slice_0->users
+  }  // end for loop operand0's slice users
+
+  // calculate the cost. If the no user found or only few small users, skip this
+  // instruction.
+  if (users.empty()) {
+    return tensorflow::errors::FailedPrecondition(
+        "No found valid users");
+  } else if (!ShouldReplace(operands, users)) {
+    return tensorflow::errors::FailedPrecondition(
+        "No Enough elements slice");
+  } else {
+    return users;
+  }
+}
+
+void SliceDelayer::GenerateNewOp(const std::vector<HloInstruction*>& operands,
+    const std::vector<HloInstruction*>& users) {
+  // generate new ops
+  const Shape shape = operands[0]->shape();
+  HloComputation* computation = users[0]->parent();
+  auto new_op = computation->AddInstruction(
+      users[0]->CloneWithNewOperands(shape, operands));
+  VLOG(10) << "Add NewOp: " << new_op->ToString();
+
+  // replace the old user and its slice operands with new operation and its
+  // slice user.
+  for (HloInstruction* user : users) {
+    const HloInstruction* slice = user->operand(0);
+    auto new_user = computation->AddInstruction(
+        slice->CloneWithNewOperands(user->shape(), {new_op}));
+    VLOG(10) << "Add NewSlice: " << new_user->ToString()
+            << "\nReplace: " << user->ToString();
```

Comment:
Don't do `\n` inside of LOGs or VLOGs.  Either put it all on one line, or do two separate log statements.  Each log message gets a header in the log file, and multiline LOGs make the log file difficult to grep.